src/LoanAgent.sol

```diff
@@ -49,73 +38,166 @@ contract LoanAgent {
         Terminated
     }
 
+    AddressInfo public addressInfo;
     NodeStatus public status;
+    LoanAgentInfo public loanAgentInfo;
+
+    /// checkpoint => withdrawnAmount
+    mapping(uint => uint) withdrawHistory;
+    /// checkpoint => available balance after withdrawal
+    mapping(uint => uint) availableBalanceHistory;
+    // checkpoint => miner share
+    mapping(uint => uint) ownerWithrawableAmount;
+    // total amount of miner share accummulated
+    uint totalOwnerCredit;
+    uint lastWithdrawlCheckpoint;
 
-    // Stakepool factory
     address public immutable factory;
+    address public immutable aFil;
 
-    // should be called using LoanAgentFactory
     constructor(
         AddressInfo memory _addressInfo,
         LoanAgentInfo memory _loanAgentInfo,
-        address _factory
+        address _factory,
+        address _aFil
     ) {
-        // Call ChangeWorkerAddress in MinerAPI: https://docs.zondax.ch/fevm/filecoin-solidity/api/actors/Miner#changeworkeraddress
+        addressInfo = _addressInfo;
+        loanAgentInfo = _loanAgentInfo;
         factory = _factory;
+        aFil = _aFil;
+        status = NodeStatus.InQueue;
     }
 
-    modifier onlyFactory {
+    modifier onlyFactory() {
+        require(msg.sender == factory, "can only be called by factory");
         _;
     }
 
-    modifier onlyOracle {
+    modifier onlyOracle() {
+        require(
+            msg.sender == addressInfo.oracle,
+            "can only be called by oracle"
+        );
         _;
     }
 
-    modifier onlyRealOwner {
+    modifier onlyOracleOrFactory() {
+        require(
+            msg.sender == addressInfo.oracle || msg.sender == factory,
+            "can only be called by oracle"
+        );
         _;
     }
 
-
-    // withdraw the funds from the Miner actor to this contract
-    function withdrawBalacne() external {}
-
-    // transfer the reward from this contract to LendingVault, Miner, Oracle, respectively
-    function distributeReward() external {}
+    modifier onlyRealOwner() {
+        require(
+            msg.sender == addressInfo.realOwner,
+            "can only be called by real owner"
+        );
+        _;
+    }
 
     // Only called by miner address
-    function changeWorker(address newWorkerAddress) external {}
+    function changeWorker(
+        address worker,
```

Comment:
No need to implement changeWorker. The protocol does not have any use.

Reply:
what if the miner wants to change worker? can it still be done if the function is not implemented here?
My understanding is that it has to be this contract address sending out the request (as it's the owner address), so it has to be implemented here. let me know if im wrong. Or you mean that we just don't allow changing worker address at all?